Source/Core/Core/IOS/USB/Emulated/Infinity.cpp

```diff
@@ -0,0 +1,958 @@
+  reply_buf[x] = GenerateChecksum(reply_buf, x);
+}
+
+void InfinityBase::GetFigureIdentifier(u8 fig_num, u8 sequence, u8* reply_buf)
+{
+  std::lock_guard lock(infinity_mutex);
+
+  InfinityFigure& figure = fig_num == m_player_one.order_added         ? m_player_one :
+                           fig_num == m_player_two.order_added         ? m_player_two :
+                           fig_num == m_player_one_ability.order_added ? m_player_one_ability :
+                           fig_num == m_player_two_ability.order_added ? m_player_two_ability :
+                                                                         m_hexagon;
+
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = 0x09;
+  reply_buf[2] = sequence;
+  reply_buf[3] = 0x00;
+
+  if (figure.present)
+  {
+    memcpy(reply_buf + 4, figure.data.data(), 7);
+  }
+  reply_buf[11] = GenerateChecksum(reply_buf, 11);
+}
+
+void InfinityBase::QueryBlock(u8 fig_num, u8 block, u8* reply_buf, u8 sequence)
+{
+  std::lock_guard lock(infinity_mutex);
+
+  InfinityFigure& figure = fig_num == m_player_one.order_added         ? m_player_one :
+                           fig_num == m_player_two.order_added         ? m_player_two :
+                           fig_num == m_player_one_ability.order_added ? m_player_one_ability :
+                           fig_num == m_player_two_ability.order_added ? m_player_two_ability :
+                                                                         m_hexagon;
+
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = 0x12;
+  reply_buf[2] = sequence;
+  reply_buf[3] = 0x00;
+  u8 file_block = 0;
+  if (block == 0)
+  {
+    file_block = 1;
+  }
+  else
+  {
+    file_block = block * 4;
+  }
+  if (figure.present)
+  {
+    memcpy(reply_buf + 4, figure.data.data() + (16 * (file_block)), 16);
+  }
+  reply_buf[20] = GenerateChecksum(reply_buf, 20);
+}
+
+void InfinityBase::WriteBlock(u8 fig_num, u8 block, const u8* to_write_buf, u8* reply_buf,
+                              u8 sequence)
+{
+  std::lock_guard lock(infinity_mutex);
+
+  InfinityFigure& figure = fig_num == m_player_one.order_added         ? m_player_one :
+                           fig_num == m_player_two.order_added         ? m_player_two :
+                           fig_num == m_player_one_ability.order_added ? m_player_one_ability :
+                           fig_num == m_player_two_ability.order_added ? m_player_two_ability :
+                                                                         m_hexagon;
+
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = 0x02;
+  reply_buf[2] = sequence;
+  reply_buf[3] = 0x00;
+  u8 file_block = 0;
+  if (block == 0)
+  {
+    file_block = 1;
+  }
+  else
+  {
+    file_block = block * 4;
+  }
+  if (figure.present)
+  {
+    memcpy(figure.data.data() + ((file_block)*16), to_write_buf, 16);
+    figure.Save();
+  }
+  reply_buf[4] = GenerateChecksum(reply_buf, 4);
+}
+
+u32 InfinityCRC32(const u8* buffer, u32 size)
+{
+  static const u32 CRC32_TABLE[256] = {
```

Comment:
Do the CRC32 functions we have in `Common/Hash.h` work for this?

Reply:
Unfortunately not, I had to generate this table manually using a separate polynomial than the ones used in the zlib functions